### ISSUE TITLE
Setting thread affinity is complicated as ManagedThreadId no longer matches ProcessThread.Id

### ISSUE THREAD

**Issue by HighPerfDotNet** [dotnet/runtime#63535]
### Description

There are instances where it is beneficial for performance to set explicit thread affinity, for some reason this functionality was not available on Thread level, so historically our .NET FX 4.x application was using the following ugly logic:

    int iProcAffinity=0;

    // Usage of GetCurrentThreadId is discouraged now as it could be volatile, but not in our config 
    // Thread.CurrentThread.ManagedThreadId can be used instead
    int iCurThreadID=AppDomain.GetCurrentThreadId();

    foreach(ProcessThread oPT in Process.GetCurrentProcess().Threads)
    {
              if(oPT.Id==iCurThreadID)
              {
                        oPT.PriorityBoostEnabled=true;
                        oPT.IdealProcessor=iProcAffinity;
                        oPT.ProcessorAffinity=new IntPtr(1L<<iProcAffinity);
                        break;
               }
    }

However in .NET 5 (possibly earlier) ProcessThread.Id no longer matched ManagedThreadId, this is a change in behavior compared to .NET FX 4.x (and earlier from at least .NET FX 2).

This is a rather old code, so if somebody knows a better way of changing current thread's processor affinity then I'd be happy with that too.

### Reproduction Steps

See code snippet above - in .NET FX current thread will be found, but not in .NET 5+

### Expected behavior

Well, ideally it should be matching what happened before, however I appreciate that there might have been good reasons for that not to be the case (and perhaps this behavior should have never been relied upon), however in such case there should be some other way to match thread using ManagedThreadId or just being able to set thread affinity right in the Thread object (but that would be more of an API proposal).

### Actual behavior

In list of all ProcessThreads their Id no longer matches ManagedThreadId 

### Regression?

Kind of - it's certainly a change, perhaps it was dangerous to rely on matching ids in the first place, but there is seemingly no easy good way to change current thread's processor affinity anyway

### Known Workarounds

I had to use ugly workaround which involved setting current managed thread priority to Lowest and then iterating through all the threads to find (hopefully) the only thread with the lowest priority, and then change thread affinity. 

### Configuration

x64, Win 10, VS 2022

**Comment by zlatanov** [comment]
There might be more more managed threads that there are OS threads and their relation is not 1 to 1. 

What I think you should be using instead is the Windows API function [GetCurrentThreadId](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getcurrentthreadid). Something like this would work:

```c#
using System.Diagnostics;
using System.Runtime.InteropServices;

var process = Process.GetCurrentProcess();
var currentThreadId = GetCurrentThreadId();

foreach (ProcessThread thread in process.Threads)
{
    if (thread.Id == currentThreadId)
    {
        // TODO
    }
}

[DllImport("kernel32.dll")]
static extern int GetCurrentThreadId();
```

**Comment by HighPerfDotNet** [comment]
Thank you for your suggestion, but I'd prefer to set processor affinity without using specific Windows API (my ugly but working alternative is using cross-platform  .NET only), it always struck me as odd that the Thread object does not have ability to set processor affinity directly, it should not be necessay to enumerate all threads to locate the one you want.

**Comment by KalleOlaviNiemitalo** [comment]
> it always struck me as odd that the Thread object does not have ability to set processor affinity directly

That's <https://github.com/dotnet/runtime/issues/21363>.

**Comment by stephentoub** [comment]
> However in .NET 5 (possibly earlier) ProcessThread.Id no longer matched ManagedThreadId, this is a change in behavior compared to .NET FX 4.x (and earlier from at least .NET FX 2).

Thread.CurrentThread.ManagedThreadId / Environment.CurrentManagedThreadId have never matched the OS thread ID; they've always been different numbering spaces, with the .NET runtime assigning ManagedThreadID and the OS assigning process thread ID.  Running this on .NET Framework 4.8:
```C#
using System;
using System.Diagnostics;
using System.Threading;

internal class Program
{
    static void Main(string[] args)
    {
        var barrier = new Barrier(20);
        for (int i = 0; i < barrier.ParticipantCount - 1; i++)
        {
            new Thread(() =>
            {
                barrier.SignalAndWait();
                Console.WriteLine($"Managed: {Thread.CurrentThread.ManagedThreadId}");
                barrier.SignalAndWait();
                barrier.SignalAndWait();
            }).Start();
        }

        barrier.SignalAndWait();
        barrier.SignalAndWait();
        foreach (ProcessThread t in Process.GetCurrentProcess().Threads)
        {
            Console.WriteLine($"ProcessThread: {t.Id}");
        }
        barrier.SignalAndWait();
    }
}
```
I get this:
```
Managed: 15
Managed: 14
Managed: 13
Managed: 19
Managed: 9
Managed: 17
Managed: 6
Managed: 7
Managed: 18
Managed: 4
Managed: 10
Managed: 20
Managed: 21
Managed: 5
Managed: 3
Managed: 8
Managed: 11
Managed: 12
Managed: 16
ProcessThread: 56444
ProcessThread: 45176
ProcessThread: 47148
ProcessThread: 45440
ProcessThread: 47268
ProcessThread: 55836
ProcessThread: 55484
ProcessThread: 55660
ProcessThread: 25496
ProcessThread: 54364
ProcessThread: 53084
ProcessThread: 49760
ProcessThread: 52924
ProcessThread: 41152
ProcessThread: 16616
ProcessThread: 50856
ProcessThread: 38880
ProcessThread: 20872
ProcessThread: 17780
ProcessThread: 54552
ProcessThread: 58260
ProcessThread: 57464
ProcessThread: 42248
ProcessThread: 4748
ProcessThread: 40048
```
Any overlap here is purely coincidental.

**Comment by zlatanov** [comment]
> Thank you for your suggestion, but I'd prefer to set processor affinity without using specific Windows API (my ugly but working alternative is using cross-platform .NET only), it always struck me as odd that the Thread object does not have ability to set processor affinity directly, it should not be necessay to enumerate all threads to locate the one you want.

The current thread affinity API is not cross platform: 
https://github.com/dotnet/runtime/blob/fd33b9eac6f259ef731e0204bcd2aebd534b8ddc/src/libraries/System.Diagnostics.Process/ref/System.Diagnostics.Process.cs#L265-L266

**Comment by HighPerfDotNet** [comment]
It works for me in .NET 4.7.2, try this -

  int iCurManagedThreadID=AppDomain.GetCurrentThreadId();
 Console.WriteLine("Current managed thead id: {0}",iCurManagedThreadID);

foreach(ProcessThread oPT in Process.GetCurrentProcess().Threads)
{
    Console.WriteLine("{0}{1}",oPT.Id,oPT.Id==iCurManagedThreadID ? " (CURRENT THREAD)" : "");
}

Output for me:

Current managed thead id: 26620
26620 (CURRENT THREAD)
5844
13168
27376
12972
28956
76
6420
8036
26584
32728
31928
30784
22292
31720
19316
20260
27388
24004
21076
26920
20920
31064
29580
11164
7700
25164
6980
20052
2500
31140

**Comment by stephentoub** [comment]
Sorry, I thought you were talking about Thread.CurrentThread.ManagedThreadId.

AppDomain.GetCurrentThreadId is deprecated / [Obsolete], even in .NET Framework:
https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.getcurrentthreadid?view=netframework-4.8
https://referencesource.microsoft.com/#mscorlib/system/appdomain.cs,2571

**Comment by HighPerfDotNet** [comment]
Thread.CurrentThread.ManagedThreadId is sadly not mapping to thread ids used by ProcessThread, but AppDomain.GetCurrentThreadId were 

The method might be obsolete, but it's behavior changed between .NET FX and new .NET, hence the report.

**Comment by stephentoub** [comment]
Yes,  this is my design. The method wasn't even included in .NET Core 1.0 because it's been obsolete/deprecated for a decade or more and the docs explicitly advise against using it due to it not being stable. When it was brought back in .NET Core 2.0 to help apps that were calling it migrate/compile, it was made to use Environment.CurrentManagedThreadId:
https://github.com/dotnet/corefx/pull/12113#discussion_r80831148

**Comment by HighPerfDotNet** [comment]
That explains for sure (even though it would have been better to keep old behavior, obsolete or not - backwards compatibility is very important), the bottom line is that a better way is needed to set thread affinity such as referenced in https://github.com/dotnet/runtime/issues/21363

**Comment by stephentoub** [comment]
> even though it would have been better to keep old behavior, obsolete or not - backwards compatibility is very important

There was never a guarantee it would consistently return a stable value for the Thread and a value that matched the OS' concept of thread id; that's why it was obsoleted.

> the bottom line is that a better way is needed to set thread affinity such as referenced in #21363

We can close this as a duplicate of that then. Thanks.

**Comment by HighPerfDotNet** [comment]
Yes, you can close it, but it's worth pointing out that https://github.com/dotnet/runtime/issues/21363 was created on 25 Apr 2017, which is almost 5 years ago

**Comment by aloraman** [comment]
> There wasn't a guarantee it would consistently return a stable value for the Thread and a value that matched the OS' concept of thread id; that's why it was obsoleted

As I recall, the mismatch was expected to happen when managed thread switches its underlying native thread (hence `Thread.BeginThreadAffinity` API exists). But in practice, that was the case only in CLR-on-SqlServer. So the wisdom of the crowd was to ignore this obsoletion warning, because "we won't be running this code on SqlServer anyway". Better solution was to reimplement this method in your own codebase (P/Invoke call to `GetCurrentThreadId` from `kernel32.dll`)

Now, current implementation of `AppDomain.GetCurrentThreadId` is weird. It's not compatible with .NetFramework , because it returns managed thread id, not native thread id. But it is stable and returns the same value as Environment.CurrentManagedThreadId, so it kinda contradicts the obsoletion warning.

IMHO, the principle of least surprise should be followed, and this method should return either the result of `Thread.GetCurrentOSThreadId` or throw an exception outright. It should not return something completely different and uncorrelated with previous behavior (see `CultureInfo.ListSeparator` debacle)

**Comment by HighPerfDotNet** [comment]
An obsolete method should act exactly as it was before or throw exception - there is zero benefit to provide managed thread id instead of previous behavior.

**Comment by KalleOlaviNiemitalo** [comment]
Adding a ProcessThread.GetCurrentThread() API (similar to Process.GetCurrentProcess()) could make some sense, too. However, it would have disadvantages:

* Calls would have to be between BeginThreadAffinity and EndThreadAffinity, for compatibility with any future green threads.
* This could not be used for setting processor affinity, ideal processor, etc. of a thread that has been created but not yet started.

**Comment by GSPP** [comment]
I recently had this situation, too. I settled on making the thread in question query its own ID using `GetCurrentThreadId` and pass that ID to other threads so that those threads could control affinity. The affinity was dynamically assigned by a form of scheduler. Each thread on its own did not have the ability to know the correct affinity mask to use. It had to be global.

It would have been easier if there had been:

```diff
public class Thread
{
+    public int NativeThreadID { get; }
}
```

In the context of possible future green threads, maybe this:

```diff
public class Thread
{
+    public int? TryGetNativeThreadID();
}
```

Being able to obtain a `ProcessThread` for a given `Thread` would be handy, too.

**Comment by ColonelBundy** [comment]
I also ran into this problem and I came up with a workaround:
```

public static class ThreadExtensions
{
    [DllImport("kernel32.dll")]
    private static extern int GetCurrentThreadId();

    [DllImport("libc.so.6")]
    private static extern int gettid();

    [SupportedOSPlatform("Windows")]
    [SupportedOSPlatform("Linux")]
    [UnsupportedOSPlatform("OSX")]
    public static int GetCurrentOsThreadId()
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
        {
            return gettid();
        }

        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
        {
            return GetCurrentThreadId();
        }

        throw new NotSupportedException("Unable to get current unmanaged thread id on this platform");
    }
}
```

**Comment by oold** [comment]
> As I recall, the mismatch was expected to happen when managed thread switches its underlying native thread (hence `Thread.BeginThreadAffinity` API exists). But in practice, that was the case only in CLR-on-SqlServer.

Not 100% accurate. I've seen that happening for Visual Studio extensions, too.

**Comment by munichmule** [comment]
@stephentoub What exactly `ProcessThread.Id` stands for then? It's neither managed thread id nor reliable os thread id. 
How can I find a ProcessThread matching my managed thread then? The suggestion by @ColonelBundy works for some win/linux scenarios, but I was unable to find a proper interop/pinvoke for macOS: ProcessThread.Id matches none of handles I can get from  `pthread_self`, `pthread_threadid_np`, `pthread_mach_thread_np` etc, so I wonder what is stands for and how it can be used/matched...
